### PR TITLE
Finished testing for Castling for both players, for all the possible …

### PR DESCRIPTION
**Pull request by EranDaniel98**
…options of threats and rules

### CODE REVIEW

**Comment by EranDaniel98** [comment]
@NitsanLitov idk why its crushing ><
in vsCode its working just fine

File: ChessTests/CastlingTest.cs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+using ChessBoard;
+using Players;
+
+namespace ChessTests
+{
+    [TestClass]
+    public class CastlingTest
+    {
+        [TestMethod]
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:14]: add categories to all the function.
movement and castling. check the other tests

File: ChessTests/CastlingTest.cs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        public void TestBlackThreatCastlingAfter()
+        {
+            Board board = TestHelper.CreateTwoPlayerBoard();
+
+            List<ChessPiece> firstPieces = board.chessPiecesByColor[PlayerColor.GetColor(PlayerNumber.FirstPlayer)];
+            List<ChessPiece> secondPieces = board.chessPiecesByColor[PlayerColor.GetColor(PlayerNumber.SecondPlayer)];
+
+            ChessPiece wKing = firstPieces[(int)FirstPiecesNumber.King];
+            ChessPiece wLeftRook = firstPieces[(int)FirstPiecesNumber.LeftRook];
+
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:87]: remove newline and do newline between white and black statements at line 91

File: ChessTests/CastlingTest.cs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+            TestHelper.ValidateMovementResults(wKing, "d1 c1");
+            TestHelper.ValidateMovementResults(bLeftBishop, "g7 h6");
+
+            //threat the space between the king and the rook
+            board.Move(('f', 8), ('g', 7));
+            TestHelper.PrintAll(board);
+            TestHelper.ValidateMovementResults(wKing, "d1 c1");
+            TestHelper.ValidateMovementResults(wLeftRook, "b1 c1 d1");
+
+            board.Move(('g', 7), ('h', 6));
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:112]: why are you moving the bishop step by step instead directly to h6?

File: ChessTests/CastlingTest.cs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            TestHelper.PrintAll(board);
+            TestHelper.ValidateMovementResults(wKing, "d1 c1");
+            TestHelper.ValidateMovementResults(wLeftRook, "b1 c1 d1");
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:115]: add bishop movement validations

File: ChessTests/CastlingTest.cs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+            bQueen.Dispose();
+            bLeftBishop.Dispose();
+            bLeftKnight.Dispose();
+            wPawn3.Dispose();
+
+            TestHelper.PrintAll(board);
+            TestHelper.ValidateMovementResults(bKing, "d8 c8");
+            TestHelper.ValidateMovementResults(bLeftRook, "b8 c8 d8");
+
+            board.Move(('d', 1), ('b', 3));
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:152]: if your'e moving the queen check her movement

File: ChessTests/CastlingTest.cs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+            TestHelper.ValidateMovementResults(bLeftRook, "b8 c8 d8");
+
+            board.Move(('b', 7), ('b', 6));
+            TestHelper.PrintAll(board);
+            TestHelper.ValidateMovementResults(bKing, "d8 c8");
+            TestHelper.ValidateMovementResults(bLeftRook, "b8 c8 d8");
+
+            board.Move(('b', 3), ('b', 6));
+            TestHelper.PrintAll(board);
+            TestHelper.ValidateMovementResults(bKing, "d8 c8");
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:164]: how can king still go to c8? shouldn't Castling be impossible?

File: ChessTests/CastlingTest.cs [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+            List<ChessPiece> secondPieces = board.chessPiecesByColor[PlayerColor.GetColor(PlayerNumber.SecondPlayer)];
+
+            ChessPiece wKing = firstPieces[(int)FirstPiecesNumber.King];
+            ChessPiece wLeftRook = firstPieces[(int)FirstPiecesNumber.LeftRook];
+
+            TestHelper.PrintAll(board);
+            TestHelper.ValidateMovementResults(wKing, "");
+            TestHelper.ValidateMovementResults(wLeftRook, "");
+
+            firstPieces[(int)FirstPiecesNumber.Queen].Dispose();
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:202]: way cleaner code. do this everywhere you dispose

Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:202]: pick a method to dispose. either do this or declare a variable like before. both have advantages

**Comment by NitsanLitov** [comment]
you're missing a threat on king directly test.

File: ChessTests/CastlingTest.cs (hunk 6 above)
Comment (NitsanLitov) [reply ChessTests/CastlingTest.cs:164]: I checked the printed result. Queen is threatning only the rook path and not the king and according to the rules thats allowed. it is a valid check thats worth checking so keep the function this way.

just add move c7 to c6 and queen will then threat king in between like you wanted.

that way youre checking both tests at onece

File: ChessTests/CastlingTest.cs (hunk 1 above)
Comment (EranDaniel98) [reply ChessTests/CastlingTest.cs:14]: done.

File: ChessTests/CastlingTest.cs (hunk 3 above)
Comment (EranDaniel98) [reply ChessTests/CastlingTest.cs:112]: you are right, i didn't notice

File: ChessTests/CastlingTest.cs (hunk 6 above)
Comment (EranDaniel98) [reply ChessTests/CastlingTest.cs:164]: Okay, I agree with you

File: ChessTests/CastlingTest.cs (hunk 7 above)
Comment (EranDaniel98) [reply ChessTests/CastlingTest.cs:202]: why does it matter so much ? 
I've changed all to be the same, but i don't understand why does it matter

File: ChessTests/CastlingTest.cs [hunk 8]
```diff
@@ -184,7 +190,44 @@ public void TestWhiteCastlingPathIsNotEmpty()
...
+        [TestMethod, TestCategory("Movement"), TestCategory("Castling")]
+        public void TestCastlingWhiteThreatBlackKing()
+        {
+            Board board = TestHelper.CreateTwoPlayerBoard();
+
+            List<ChessPiece> firstPieces = board.chessPiecesByColor[PlayerColor.GetColor(PlayerNumber.FirstPlayer)];
+            List<ChessPiece> secondPieces = board.chessPiecesByColor[PlayerColor.GetColor(PlayerNumber.SecondPlayer)];
+
+            ChessPiece bKing = secondPieces[(int)SecondPiecesNumber.King];
+            ChessPiece bRightRook = secondPieces[(int)SecondPiecesNumber.LeftRook];
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:202]: its not right its left

File: ChessTests/CastlingTest.cs [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+            ChessPiece wKing = firstPieces[(int)FirstPiecesNumber.King];
+            ChessPiece wRightRook = firstPieces[(int)FirstPiecesNumber.RightRook];
+
+            ChessPiece wRightBishop = firstPieces[(int)FirstPiecesNumber.RightBishop];
+            ChessPiece wRightKnight = firstPieces[(int)FirstPiecesNumber.RightKnight];
+
+            TestHelper.PrintAll(board);
+            TestHelper.ValidateMovementResults(wKing, "");
+
+            wRightKnight.Dispose();
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:29]: ?

File: ChessTests/CastlingTest.cs [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+            wRightBishop.Dispose();
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:30]: ?

File: ChessTests/CastlingTest.cs [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+
+            ChessPiece bRightRook = secondPieces[(int)SecondPiecesNumber.RightRook];
+            ChessPiece bRightBishop = secondPieces[(int)SecondPiecesNumber.RightBishop];
+            ChessPiece bRightKnight = secondPieces[(int)SecondPiecesNumber.RightKnight];
+
+            TestHelper.PrintAll(board);
+            TestHelper.ValidateMovementResults(bKing, "");
+            TestHelper.ValidateMovementResults(bQueen, "");
+
+            bRightBishop.Dispose();
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:61]: ?

File: ChessTests/CastlingTest.cs [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+            bRightKnight.Dispose();
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:62]: ?

File: ChessTests/CastlingTest.cs [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+            bQueen.Dispose();
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:63]: ?

File: ChessTests/CastlingTest.cs [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+            List<ChessPiece> firstPieces = board.chessPiecesByColor[PlayerColor.GetColor(PlayerNumber.FirstPlayer)];
+            List<ChessPiece> secondPieces = board.chessPiecesByColor[PlayerColor.GetColor(PlayerNumber.SecondPlayer)];
+
+            ChessPiece wKing = firstPieces[(int)FirstPiecesNumber.King];
+            ChessPiece wLeftRook = firstPieces[(int)FirstPiecesNumber.LeftRook];
+
+            ChessPiece bLeftBishop = secondPieces[(int)SecondPiecesNumber.LeftBishop];
+
+            TestHelper.PrintAll(board);
+            TestHelper.ValidateMovementResults(wKing, "");
```
Comment (NitsanLitov) [inline ChessTests/CastlingTest.cs:90]: add new line

**Review by NitsanLitov — APPROVED** [review]

**Pull request merged**